File: frontend/src/api/ragicFieldAudit.ts
/**
 * Ragic 與 Portal 欄位比對 API
 * 路由前綴：/api/v1/settings/ragic-field-audit
 */
import apiClient from '@/api/client'

const BASE = '/settings/ragic-field-audit'

// ── 型別定義 ──────────────────────────────────────────────────────────────────

export interface AuditSummary {
  total_modules: number
  audited_modules: number
  normal_modules: number
  error_modules: number
  unmapped_fields: number
  high_risk_issues: number
  last_run_time: string | null
  last_run_status: string | null
}

export interface ModuleOverview {
  item_no: number
  company: string
  module_name: string
  portal_name: string
  portal_route: string
  local_tables: string[]
  ragic_url: string
  ragic_field_count: number
  portal_db_field_count: number
  portal_api_field_count: number
  portal_fe_field_count: number
  normal_count: number
  issue_count: number
  unmapped_count: number
  total_mapping_count: number
  last_checked_at: string | null
  status: 'normal' | 'warning' | 'error' | 'not_audited'
  is_active: boolean
}

export interface FieldMapping {
  id: number | null
  app_directory_id: number | null
  company: string | null
  module_name: string | null
  portal_route: string | null
  ragic_url: string | null
  ragic_form_name: string | null
  ragic_field_id: string
  ragic_field_name: string
  ragic_field_type: string
  is_ragic_required: boolean
  is_ragic_formula: boolean
  is_ragic_subtable: boolean
  portal_db_table: string
  portal_db_field: string
  portal_api_field: string
  portal_frontend_field: string
  display_name: string
  is_displayed: boolean
  is_filter: boolean
  is_export: boolean
  is_calculated: boolean
  mapping_status: string
  severity: string | null
  issue_type: string | null
  issue_message: string | null
  suggestion: string | null
  is_resolved: boolean
  notes: string | null
  last_checked_at: string | null
  // 草稿模式（API 生成但未存 DB）
  category?: string
}

export interface KpiMapping {
  id: number | null
  module_name: string | null
  portal_route: string | null
  kpi_name: string
  page_section: string | null
  api_endpoint: string | null
  db_table: string | null
  source_fields: string | null
  date_field: string | null
  filters: string | null
  formula: string | null
  ragic_source_fields: string | null
  trace_status: 'traceable' | 'partial' | 'untraceable' | 'unknown'
  issue_message: string | null
  suggestion: string | null
  last_checked_at: string | null
}

export interface AuditRun {
  id: number
  run_time: string
  triggered_by: string | null
  scope: string | null
  total_modules: number
  normal_count: number
  warning_count: number
  error_count: number
  status: string
  notes: string | null
}

// ── API 函數 ──────────────────────────────────────────────────────────────────

/** 首頁 KPI Card 摘要 */
export async function fetchAuditSummary(): Promise<AuditSummary> {
  const res = await apiClient.get(`${BASE}/summary`)
  return res.data
}

/** 模組比對總覽（Tab 1） */
export async function fetchModules(params?: {
  company?: string
  status?: string
  keyword?: string
}): Promise<{ items: ModuleOverview[]; total: number }> {
  const res = await apiClient.get(`${BASE}/modules`, { params })
  return res.data
}

/** 單一模組欄位 Mapping 明細（Tab 2） */
export async function fetchModuleDetail(route: string): Promise<{
  route: string
  items: FieldMapping[]
  total: number
}> {
  const res = await apiClient.get(`${BASE}/module`, { params: { route } })
  return res.data
}

/** 異常清單（Tab 3） */
export async function fetchIssues(params?: {
  severity?: string
  is_resolved?: boolean
  module_name?: string
  keyword?: string
}): Promise<{ items: FieldMapping[]; total: number }> {
  const res = await apiClient.get(`${BASE}/issues`, { params })
  return res.data
}

/** KPI 計算追溯（Tab 4） */
export async function fetchKpiMappings(params?: {
  module_name?: string
  trace_status?: string
}): Promise<{ items: KpiMapping[]; total: number }> {
  const res = await apiClient.get(`${BASE}/kpi-mappings`, { params })
  return res.data
}

/** 執行比對稽核 */
export async function runAudit(scope = 'all'): Promise<{
  run_id: number
  run_time: string
  total_modules: number
  normal_count: number
  warning_count: number
  error_count: number
  created_mappings: number
  status: string
}> {
  const res = await apiClient.post(`${BASE}/run`, { scope })
  return res.data
}

/** 匯出 Excel（觸發下載） */
export async function exportExcel(): Promise<void> {
  const res = await apiClient.get(`${BASE}/export`, { responseType: 'blob' })
  const url = URL.createObjectURL(new Blob([res.data]))
  const a = document.createElement('a')
  a.href = url
  const now = new Date().toISOString().slice(0, 19).replace(/[-:T]/g, (c) =>
    c === 'T' ? '_' : c === ':' ? '' : c
  )
  a.download = `ragic_field_audit_${now}.xlsx`
  a.click()
  URL.revokeObjectURL(url)
}

/** 標記異常為已處理 */
export async function resolveMappingIssue(
  mappingId: number,
  isResolved: boolean,
  notes?: string,
): Promise<{ id: number; is_resolved: boolean; notes: string | null }> {
  const res = await apiClient.patch(`${BASE}/mapping/${mappingId}/resolve`, {
    is_resolved: isResolved,
    notes,
  })
  return res.data
}

/** 歷史比對執行紀錄 */
export async function fetchAuditRuns(): Promise<{ items: AuditRun[] }> {
  const res = await apiClient.get(`${BASE}/runs`)
  return res.data
}

export interface SyncRagicFieldsResult {
  item_no: number
  ragic_url: string
  portal_route: string
  ragic_field_count: number
  synced_count: number
  updated_count: number
  ragic_fields: string[]
  fetch_error: string | null
}

/** 從 Ragic API 抓取指定表單的所有欄位並儲存至比對表 */
export async function syncRagicFields(
  itemNo: number,
  ragicUrl: string,
): Promise<SyncRagicFieldsResult> {
  const res = await apiClient.post(`${BASE}/sync-ragic-fields`, {
    item_no: itemNo,
    ragic_url: ragicUrl,
  })
  return res.data
}

/** 取得系統已知的 Ragic URL 對照表（itemNo → ragic_url） */
export async function fetchRagicUrlMap(): Promise<{ items: { item_no: number; ragic_url: string }[] }> {
  const res = await apiClient.get(`${BASE}/ragic-url-map`)
  return res.data
}

/** 設定（或清除）指定模組的 Ragic 表單 URL */
export async function setModuleRagicUrl(
  itemNo: number,
  ragicUrl: string,
): Promise<{ item_no: number; ragic_url: string; portal_route: string; updated_mappings: number }> {
  const res = await apiClient.patch(`${BASE}/modules/${itemNo}/ragic-url`, {
    ragic_url: ragicUrl,
  })
  return res.data
}

File: backend/app/models/ragic_field_audit.py
"""
Ragic 與 Portal 欄位比對 — 資料模型

包含三張表：
  1. ragic_portal_audit_runs      — 比對任務紀錄（每次執行比對的快照）
  2. ragic_portal_field_mappings  — 欄位對應關係（Ragic ↔ Portal DB / API / 前端）
  3. ragic_portal_kpi_mappings    — KPI / Dashboard 計算來源追溯
"""

from datetime import datetime
from typing import Optional
from sqlalchemy import Integer, String, Text, DateTime, Boolean, Float
from sqlalchemy.orm import mapped_column, Mapped
from app.core.database import Base
from app.core.time import twnow


class RagicPortalAuditRun(Base):
    """記錄每次欄位比對任務的執行結果摘要。"""
    __tablename__ = "ragic_portal_audit_runs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    run_time: Mapped[datetime] = mapped_column(DateTime, default=twnow)
    triggered_by: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)  # 使用者 email
    scope: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)         # "all" 或 module_name
    total_modules: Mapped[int] = mapped_column(Integer, default=0)
    normal_count: Mapped[int] = mapped_column(Integer, default=0)
    warning_count: Mapped[int] = mapped_column(Integer, default=0)
    error_count: Mapped[int] = mapped_column(Integer, default=0)
    status: Mapped[str] = mapped_column(String(20), default="completed")             # running / completed / failed
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)


class RagicPortalFieldMapping(Base):
    """
    記錄 Ragic 欄位與 Portal 各層欄位的對應關係。
    每筆代表一個模組的一個欄位對應紀錄。
    """
    __tablename__ = "ragic_portal_field_mappings"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    # ── 模組識別 ──────────────────────────────────────────────────────────────
    app_directory_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)   # ragic_app_portal_annotations.item_no
    company: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)        # 公司 / 據點
    module_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)    # Portal 模組中文名稱
    portal_route: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)   # /settings/ragic-field-audit
    ragic_url: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    ragic_form_name: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)

    # ── Ragic 欄位資訊 ────────────────────────────────────────────────────────
    ragic_field_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    ragic_field_name: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)
    ragic_field_type: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)  # text / number / date / select / formula / subtable
    is_ragic_required: Mapped[bool] = mapped_column(Boolean, default=False)
    is_ragic_formula: Mapped[bool] = mapped_column(Boolean, default=False)
    is_ragic_subtable: Mapped[bool] = mapped_column(Boolean, default=False)
    ragic_options: Mapped[Optional[str]] = mapped_column(Text, nullable=True)          # JSON array of select options

    # ── Portal 各層欄位 ───────────────────────────────────────────────────────
    portal_db_table: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    portal_db_field: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    portal_api_field: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    portal_frontend_field: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    display_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)   # 前端中文顯示名稱

    # ── 使用情況標記 ──────────────────────────────────────────────────────────
    is_displayed: Mapped[bool] = mapped_column(Boolean, default=False)    # 有出現在頁面
    is_filter: Mapped[bool] = mapped_column(Boolean, default=False)       # 用於篩選
    is_export: Mapped[bool] = mapped_column(Boolean, default=False)       # 用於匯出
    is_calculated: Mapped[bool] = mapped_column(Boolean, default=False)   # 用於計算

    # ── 比對結果 ──────────────────────────────────────────────────────────────
    # mapping_status 可能值：
    #   normal          — 正常對應
    #   ragic_only      — Ragic 有，Portal 沒有
    #   portal_only     — Portal 有，Ragic 沒有
    #   name_mismatch   — 欄位名稱疑似不同（fuzzy match）
    #   type_mismatch   — 型態不一致
    #   null_rate_high  — 空值率異常
    #   formula_unmarked — 公式欄位未標示
    #   subtable_unmarked — 子表格欄位未處理
    #   unmapped        — 未建立 Mapping
    mapping_status: Mapped[str] = mapped_column(String(50), default="unmapped")
    severity: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)        # high / medium / low
    issue_type: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    issue_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)         # 主管看得懂的說明
    suggestion: Mapped[Optional[str]] = mapped_column(Text, nullable=True)            # 建議處理方式
    is_resolved: Mapped[bool] = mapped_column(Boolean, default=False)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    last_checked_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=twnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=twnow, onupdate=twnow)


class RagicPortalKpiMapping(Base):
    """
    記錄 Dashboard / 統計模組的 KPI 與 Ragic 原始欄位的可追溯關係。
    """
    __tablename__ = "ragic_portal_kpi_mappings"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    # ── 模組識別 ──────────────────────────────────────────────────────────────
    app_directory_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    module_name: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    portal_route: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)

    # ── KPI 描述 ──────────────────────────────────────────────────────────────
    kpi_name: Mapped[str] = mapped_column(String(200))                               # KPI 名稱
    page_section: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)  # 顯示位置（頁面區塊）

    # ── 資料來源 ──────────────────────────────────────────────────────────────
    api_endpoint: Mapped[Optional[str]] = mapped_column(String(300), nullable=True)
    db_table: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    source_fields: Mapped[Optional[str]] = mapped_column(Text, nullable=True)        # JSON array，使用的欄位
    date_field: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)    # 日期依據欄位
    filters: Mapped[Optional[str]] = mapped_column(Text, nullable=True)              # JSON：篩選條件
    formula: Mapped[Optional[str]] = mapped_column(Text, nullable=True)              # 計算公式說明

    # ── Ragic 可追溯性 ───────────────────────────────────────────────────────
    ragic_source_fields: Mapped[Optional[str]] = mapped_column(Text, nullable=True)  # JSON array
    # trace_status 可能值：traceable / partial / untraceable / unknown
    trace_status: Mapped[str] = mapped_column(String(30), default="unknown")
    issue_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    suggestion: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    last_checked_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=twnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=twnow, onupdate=twnow)
